Fixes #29145 - Add custom evr type column to katello_rpms and katello_installed_packages

ianballou opened the pull request:
Includes custom sql types, functions, and triggers that automatically generate items in the new "evr" column for katello_rpms and katello_installed_packages. The migrations that add these types are all reversible. The logic for creating the evr type is mostly pulled from here: https://github.com/ggainey/pulp_startup/blob/master/applicability/evr_t.sql. The evr type is treated as a string by ActiveRecord, but this shouldn't be an issue since the EVR type shouldn't be used directly in Ruby code.

To test: 
1) Ensure there are some RPMs and InstalledPackages in your database.  
2) Install this RPM: https://fedorapeople.org/groups/katello/releases/yum/nightly/pulpcore/el7/x86_64/rh-postgresql12-postgresql-evr-0.0.1-2.el7.x86_64.rpm
2) Checkout my branch and run the migration (ignore the new warning about the evr_t type). 
3) Check that the existing RPMs and InstalledPackages have something other than `nil` for evr.  
4) Sync in or create some RPMs and see that the evr column for katello_rpms is updated. 
5) Create some InstalledPackages and see that evr is updated too.  The evr column will show up as a weird string, but that's okay since it'll be used in the background. 
6) Check that the evr columns are updated if the epochs, versions, or releases are updated.
7) Try rolling back the migration.

TODO:

- [x] Get postgresql-evr package into foreman-packaging (https://github.com/theforeman/foreman-packaging/pull/4795)
- [x] Get postgresql-evr installed on the Jenkins slaves

ianballou commented:
Something is up with these custom types and `schema.rb`.  I see this comment: `Could not dump table "katello_installed_packages" because of following StandardError Unknown type 'evr_t' for column 'evr'`

ianballou commented:
Looking like I might need to patch the postgres AR adapter to make the evr_t type recognized.  Either that or use this:
```# Use SQL instead of Active Record's schema dumper when creating the database.
# This is necessary if your schema can't be completely dumped by the schema dumper,
# like if you have constraints or database-specific column types
config.active_record.schema_format = :sql
```

Switching the schema foreman to SQL might be a big deal granted Foreman + Katello's complexity.

ianballou commented:
Patching in the `evr_t` into the Active Record PostgreSQL adapter has removed the unknown type error and the issues with `db:test:load_schema`.  Now I need to see how I can have those patches exist in the Katello code base.

ianballou commented:
The PostgreSQL Adapter patching is a WIP right now.

ianballou commented:
Looks like the failing `db:test:load_schema` is running on a pre-migrated database that doesn't have the `evr_t` type.  Need to find a way to load in `evr_t` earlier I think.

ianballou commented:
To avoid all these troubles, I'm going to look into creating a postgres extension to have the `evr_t` type enter the database much earlier.

ianballou commented:
Creating the extension works fine.  The next step is to have the extension packaged into an RPM to make installation easy with the foreman installer.

To add the `evr_t` type, `CREATE EXTENSION evr_t` will need to be run ideally immediately after `katello` and `katello_test` databases are created.

ianballou commented:
foreman-packaging PR for the evr-postgresql RPM: https://github.com/theforeman/foreman-packaging/pull/4795

ianballou commented:
The monkey patching fixed issues with `load_schema` but not with the warning message that occurs once on startup.  This shouldn't become a blocking issue so we can make a redmine for the remaining warning message if it's not fixed by the time this PR is merged.

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

johnpmitsch commented on test/models/rpm_test.rb:
```diff
@@ -404,6 +404,12 @@ def test_search_compare_lt_gte
+      rpm = Rpm.where(nvra: "abc123-2-1.0.0-1.0").first
+
+      assert_equal "(1,\"{\"\"(1,)\"\",\"\"(0,)\"\",\"\"(0,)\"\"}\",\"{\"\"(1,)\"\",\"\"(0,)\"\"}\")", rpm.evr
+    end
```
Could you test updating an attribute on a record and getting back an appropriate updated evr string?

johnpmitsch replied:
ideally there is a test for each of the attributes that affects the evr string

ianballou commented:
@johnpmitsch I updated the tests, let me know what you think.

ianballou commented:
Looking into why the tests failed.  Also this [PR](https://github.com/theforeman/puppet-katello_devel/pull/230) plus another PR for nightly katello needs to be added to make sure that they get bundled with the EVR package.

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

jlsherrill commented:
strange that https://github.com/Katello/katello/pull/8621 passed, are you sure it had the same exact code?

ianballou commented:
> strange that #8621 passed, are you sure it had the same exact code?

I've rebased my other PR on top of this one and it's passing locally, not sure why it's failing. I've rebased and pushed again.

ianballou commented:
[test katello]

ianballou commented:
I'm pretty stumped as to why that test is failing, even the next evr tests are passing.

ianballou commented:
Closing since #8621 included these changes.
